pkg/reconciler/tenancy/clusterworkspacedeletion/clusterworkspace_deletion_controller.go

```diff
@@ -242,6 +229,31 @@ func (c *Controller) patchCondition(ctx context.Context, old, new *tenancyv1alph
 	return err
 }
 
+func (c *Controller) patchFinalizer(ctx context.Context, ws *tenancyv1alpha1.ClusterWorkspace, finalizers []string) error {
+	finalizerData := &metav1.PartialObjectMetadata{
+		ObjectMeta: metav1.ObjectMeta{
+			ResourceVersion: ws.ResourceVersion,
+			UID:             ws.UID,
+			Finalizers:      finalizers,
+		},
+	}
+
+	patch, err := json.Marshal(finalizerData)
+	if err != nil {
+		return err
+	}
+
+	// remove finalizers field if there is no remaining finalizers,
+	if len(finalizers) == 0 {
+		patch = []byte(
+			fmt.Sprintf("{\"metadata\": {\"finalizers\": [], \"uid\": %q, \"resourceVersion\": %q}}", ws.UID, ws.ResourceVersion))
```

Comment:
At this point it's much easier to reason about a simple Update call with `ws` from above with the changes directly in the object.

Reply:
updated